Add support for icm42686 and tune for NXP VMURT1170

sumitbatra-nxp opened the pull request:
1. Support both icm42688 and icm42686 through the same driver
2. Add support for setting Anti Alias Filter for Accelerometer and Gyroscope
3. Tune icm4268x sensors for VMURT1170 for NXP's Mobile Robotics
   products

decsny commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,318 @@
+      - 1 # ICM4268X_DT_TMST_RES_RTC
+
+  tmst-regs-en:
+    type: int
```
boolean

decsny commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,318 @@
+      If 0, timestamp resolution is 1us, if 1 and RTC is enabled then time res is
+      RTC clock period else 16us.
+    enum:
+      - 0 # ICM4268X_DT_TMST_RES_1US
```
consider as string enum for these types of things

decsny commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,318 @@
+
+  int1-pol:
+    type: int
+    default: 0
```
looks like this should not have a default but maybe be required: true. otherwise, justify the default

decsny commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,318 @@
+  int-gpios:
+    type: phandle-array
+    description: |
+      The INT signal default configuration is active-high. The
```
nit -> say what this signal is in the description (even if it is obvious)

decsny commented on include/zephyr/dt-bindings/sensor/icm4268x.h:
since you have all these #defines, and if you expect them to be used somewhere other than devicetree, then per my other comment maybe lean toward putting them in the description rather than string enums

sumitbatra-nxp replied:
@decsny .. Can you please check now

sumitbatra-nxp replied:
@decsny .. please check now

decsny commented on samples/sensor/icm4268x/prj.conf:
```diff
@@ -0,0 +1,13 @@
+
+# Debug options.
+CONFIG_LOG=y
+CONFIG_SENSOR_LOG_LEVEL_DBG=y
```
why debug level logs by default for the sample?

decsny commented on samples/sensor/icm4268x/src/main.c:
```diff
@@ -0,0 +1,86 @@
+
+int main(void)
+{
+	const struct device *const icm42688 = DEVICE_DT_GET_ONE(invensense_icm42688);
```
perhaps use an alias? not required.

sumitbatra-nxp replied:
Seeing - samples/sensor/icm42605/src/main.c.. I feel let it be

sumitbatra-nxp commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,318 @@
+
+  fifo-mode:
+    type: int
+    default: 3
```
Will it work if I mention following in the description -
"Setting the default value as per the datasheet section number XXX"

decsny replied:
I think, if the datasheet recommends a default value for some reason, you should put that reason here. If it is a reset value then just say that.

sumitbatra-nxp commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,318 @@
+    description: |
+      Specify the Anti Aliasing Filter bandwidth for the accelerometer.
+    enum:
+      - 1  # ICM4268X_DT_ACCEL_AAF_42HZ
```
I followed an example like - dts/bindings/sensor/ti,tmag5273.yaml
So is it ok if I mention about the DT header file the description like how this file mentions -

    **When setting the enum properties in a .dts or .dtsi file you may
    include tmag5273.h and use the macros defined there.

    Example:
    #include <zephyr/dt-bindings/sensor/tmag5273.h>

    tmag5273: tmag5273@0 {
      ...
      axis = <TMAG5273_DT_AXIS_XYZ>;
      range = <TMAG5273_DT_AXIS_RANGE_HIGH>;
    };**

sumitbatra-nxp replied:
@decsny .. Please check now

teburd commented:
Please split up the first commit into a few commits...

1. renaming icm42688 to icm4268x and adding the additional 686 compatible
2. Adding additional dt config options
3. Adding sample app (I'm not entirely convinced of this? whats special about this sample?)

The last commit modifying the vmu board dts seems ok to me.

sumitbatra-nxp commented:
> Please split up the first commit into a few commits...
> 
> 1. renaming icm42688 to icm4268x and adding the additional 686 compatible
> 2. Adding additional dt config options
> 3. Adding sample app (I'm not entirely convinced of this? whats special about this sample?)
> 
> The last commit modifying the vmu board dts seems ok to me.
@teburd 
There is no real use of the sample app, I just wrote it for my testing and kept it as is.
Please suggest if you want me to remove it.

sumitbatra-nxp commented:
> Please split up the first commit into a few commits...
> 
> 1. renaming icm42688 to icm4268x and adding the additional 686 compatible
> 2. Adding additional dt config options
> 3. Adding sample app (I'm not entirely convinced of this? whats special about this sample?)
> 
> The last commit modifying the vmu board dts seems ok to me.

@teburd .. Created the commits as per your suggestion.. Please check now

decsny commented on dts/bindings/sensor/invensense,icm42688.yaml:
```diff
@@ -4,111 +4,48 @@
-      - 10 # ICM42688_DT_GYRO_ODR_25
-      - 11 # ICM42688_DT_GYRO_ODR_12_5
-      - 15 # ICM42688_DT_GYRO_ODR_500
+      - 0 # ICM42688_DT_ACCEL_FS_16
```
these comments on all the properties are still not actually going into the documentation if they are not part of the description

sumitbatra-nxp replied:
@decsny .. check now

decsny commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,356 @@
+      Both 0 and 1 values correspond to the same setting.
+    enum:
+      - 1 # ICM4268X_DT_FIFO_THS_INT_CLR_STSRD
+      - 2 # ICM4268X_DT_FIFO_THS_INT_CLR_DATARD
```
these are still comments, please update the description

decsny commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,356 @@
+    default: 1
+    description: |
+      Fifo threshold Interrupt Clearing Options.
+      Default value is set as per the datasheet section 14.49.
```
all of these "default value set as per the datasheet" is not an actual explanation, it is something, and it can still mention the datasheet section, but also need to clearly say what is the reason for the value in this description, not just make the user go to investigate on their own. The point of the descriptions is to be descriptive.

sumitbatra-nxp replied:
@decsny .. Changed.. please check

decsny commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,356 @@
+  tmst-regs-en:
+    type: boolean
+    description: |
+      If zero, TMST_VALUE returns zeros else returns timestamp values.
```
if false

decsny commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,368 @@
+    default: 6
+    description: |
+      Specify the accelerometer output data rate expressed in samples per second (Hz).
+      Default accelerometer ODR is 1kHz in Low Noise Mode.
```
I feel like there is still a misunderstanding here about the default values, please read this link https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values, you need to say *why* something is the default value

sumitbatra-nxp replied:
The reason for choosing a default value is not mentioned in the Invensense datasheet. 
Though I can think of a logical reason myself.. but is it recommended to write something on behalf of the part owner ?

decsny replied:
> The reason for choosing a default value is not mentioned in the Invensense datasheet. Though I can think of a logical reason myself.. but is it recommended to write something on behalf of the part owner ? Also this link for Zephyr default rules has an example where it says - "The default was recommended in the manufacturer datasheet"
> 
> So if you recommend, I can write like this - "The default value for accelerometer ODR as recommended in the manufacturer datasheet is 1kHz in Low Noise Mode"

that's fine to write that

sumitbatra-nxp replied:
Done.. please check

teburd commented:
Will review once CI passes

decsny commented on dts/bindings/sensor/invensense,icm4268x-common.yaml:
```diff
@@ -0,0 +1,368 @@
+    type: boolean
+    description: |
+      Enable or disable the timestamp register.
+      Timestamp register is enabled by default.
```
The default of a boolean is false....

sumitbatra-nxp replied:
Changed the parameter name to tmst-dis instead.. please check

sumitbatra-nxp committed:
sensor: tdk: Add support for icm42686

Support both icm42686 and icm42688 through the same driver

Co-authored-by: Mayank Mahajan <mayankmahajan.x@nxp.com>
Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

sumitbatra-nxp committed:
sensor: tdk: Additional dt options for icm4268x

Add Filters for Accelerometer and Gyroscope and other dt options

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

sumitbatra-nxp committed:
boards: nxp: Tune icm4286x sensors for VMURT7

Populate vmu_rt1170 board dts file with default settings for icm42688
and icm42686.

Co-authored-by: Mayank Mahajan <mayankmahajan.x@nxp.com>
Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

sumitbatra-nxp commented:
@teburd .. Please review..

teburd commented:
@bperseghetti do take a look if you can

teburd commented:
Marking DNM until people have a chance to test this out, as it is a pretty big change. I'll try and test it out next week.

sumitbatra-nxp commented:
@teburd .. The source repo is deleted due to some internal policy related issues.. but please continue your testing.. I'll raise it from another Repo and link it to this one
